modules/product/src/main/java/com/opengamma/strata/product/fx/FxSingleBarrierOption.java

```diff
@@ -0,0 +1,452 @@
+/**
+ * Copyright (C) 2016 - present by OpenGamma Inc. and the OpenGamma group of companies
+ *
+ * Please see distribution for license.
+ */
+package com.opengamma.strata.product.fx;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.Set;
+
+import org.joda.beans.Bean;
+import org.joda.beans.BeanBuilder;
+import org.joda.beans.BeanDefinition;
+import org.joda.beans.ImmutableBean;
+import org.joda.beans.ImmutableValidator;
+import org.joda.beans.JodaBeanUtils;
+import org.joda.beans.MetaProperty;
+import org.joda.beans.Property;
+import org.joda.beans.PropertyDefinition;
+import org.joda.beans.impl.direct.DirectFieldsBeanBuilder;
+import org.joda.beans.impl.direct.DirectMetaBean;
+import org.joda.beans.impl.direct.DirectMetaProperty;
+import org.joda.beans.impl.direct.DirectMetaPropertyMap;
+
+import com.opengamma.strata.basics.currency.CurrencyAmount;
+import com.opengamma.strata.basics.market.ReferenceData;
+import com.opengamma.strata.basics.market.Resolvable;
+import com.opengamma.strata.collect.ArgChecker;
+import com.opengamma.strata.product.Product;
+
+/**
+ * FX (European) single barrier option. 
+ * <p>
+ * An FX option is a financial instrument that provides an option to exchange two currencies at a specified future time 
+ * only when barrier event occurs (knock-in option) or does not occur (knock-out option).
+ * <p>
+ * Depending on the barrier defined in {@link Barrier}, the options are classified into four types: up-and-in, 
+ * down-and-in, up-and-out and down-and-out.
+ * <p>
+ * For example, an up-and-out call on a 'EUR 1.00 / USD -1.41' exchange with barrier of 1.5 is the option to
+ * perform a foreign exchange on the expiry date, where USD 1.41 is paid to receive EUR 1.00, only when EUR/USD rate does 
+ * not exceed 1.5 during the barrier event observation period. 
+ * <p>
+ * In case of the occurrence (non-occurrence for knock-in options) of the barrier event, the option becomes worthless, 
+ * or alternatively, a rebate is made. 
+ */
+@BeanDefinition(builderScope = "private")
+public final class FxSingleBarrierOption
+    implements Product, Resolvable<ResolvedFxSingleBarrierOption>, ImmutableBean, Serializable {
+
+  /**
+   * The underlying FX vanilla option.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final FxVanillaOption underlyingOption;
+  /**
+   * The barrier description.
+   * <p>
+   * The barrier level stored in this field must be represented based on the direction of the currency pair in the 
+   * underlying FX transaction. 
+   * <p>
+   * For example, if the underlying option is an option on EUR/GBP, the barrier should be a certain level of EUR/GBP rate.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final Barrier barrier;
+  /**
+   * The amount paid back to the option holder in case the option becomes inactive.
```

Comment:
For a 'out' option, the amount is paid when the barrier is reached; for a 'in' option, the amount is paid at expiry if the barrier is not reached.